crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2026,6 +2028,118 @@ impl<'db> Class<'db> {
         self.iter_mro(db).contains(&ClassBase::Class(other))
     }
 
+    /// Return the explicit `metaclass` of this class, if one is defined.
+    ///
+    /// ## Note
+    /// Only call this function from queries in the same file or your
+    /// query depends on the AST of another file (bad!).
+    fn explicit_metaclass(self, db: &'db dyn Db) -> Option<Type<'db>> {
+        let class_stmt = self.node(db);
+        let metaclass = class_stmt
+            .arguments
+            .as_ref()?
+            .find_keyword("metaclass")
+            .map(|keyword| &keyword.value);
+        if class_stmt.type_params.is_some() {
+            // when we have a specialized scope, we'll look up the inference
+            // within that scope
+            let model = SemanticModel::new(db, self.file(db));
+            metaclass.map(|base| base.ty(&model))
+        } else {
+            // Otherwise, we can do the lookup based on the definition scope
+            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
+            metaclass.map(|base_node| definition_expression_ty(db, class_definition, base_node))
+        }
```

Comment:
```suggestion
        let metaclass_node = &class_stmt
            .arguments
            .as_ref()?
            .find_keyword("metaclass")?
            .value;
        if class_stmt.type_params.is_some() {
            // when we have a specialized scope, we'll look up the inference
            // within that scope
            let model = SemanticModel::new(db, self.file(db));
            metaclass_node.ty(&model)
        } else {
            // Otherwise, we can do the lookup based on the definition scope
            let class_definition = semantic_index(db, self.file(db)).definition(class_stmt);
            definition_expression_ty(db, class_definition, metaclass_node)
        }
```